pkg/controller/traceflow/validate_test.go

```diff
@@ -0,0 +1,175 @@
+		name string
+
+		// environment
+		pods []v1.Pod
+
+		// input
+		oldSpec *crdv1alpha1.TraceflowSpec
+		newSpec *crdv1alpha1.TraceflowSpec
+
+		// expected output
+		allowed      bool
+		deniedReason string
+	}{
+		{
+			name: "Validating destination IP",
+			newSpec: &crdv1alpha1.TraceflowSpec{
+				Destination: crdv1alpha1.Destination{IP: "foobar"},
+			},
+			deniedReason: "destination IP is not valid: foobar",
+		},
+		{
+			name: "Traceflow should have either source or destination Pod assigned",
+			newSpec: &crdv1alpha1.TraceflowSpec{
+				LiveTraffic: true,
+			},
+			deniedReason: "Traceflow tf has neither source nor destination Pod specified",
+		},
+		{
+			name: "Assigned source pod must exist",
+			newSpec: &crdv1alpha1.TraceflowSpec{
+				Source: crdv1alpha1.Source{
+					Namespace: "test-ns",
+					Pod:       "test-pod",
+				},
+			},
+			deniedReason: "requested source Pod test-ns/test-pod not found",
+		},
+		{
+			name: "Using hostNetwork Pod as source in non-live-traffic Traceflow is not supported",
+			pods: []v1.Pod{
+				{
+					ObjectMeta: metav1.ObjectMeta{Namespace: "test-ns", Name: "test-pod"},
+					Spec:       v1.PodSpec{HostNetwork: true},
+				},
+			},
+			newSpec: &crdv1alpha1.TraceflowSpec{
+				Source: crdv1alpha1.Source{
+					Namespace: "test-ns",
+					Pod:       "test-pod",
+				},
+			},
+			deniedReason: "using hostNetwork Pod as source in non-live-traffic Traceflow is not supported",
+		},
+		{
+			name: "Valid request",
+			pods: []v1.Pod{
+				{
+					ObjectMeta: metav1.ObjectMeta{Namespace: "test-ns", Name: "test-pod"},
+				},
+			},
+			newSpec: &crdv1alpha1.TraceflowSpec{
+				Source: crdv1alpha1.Source{
+					Namespace: "test-ns",
+					Pod:       "test-pod",
+				},
+			},
+			allowed: true,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			stopCh := make(chan struct{})
+			defer close(stopCh)
+			controller := newController()
+			controller.informerFactory.Start(stopCh)
+			controller.crdInformerFactory.Start(stopCh)
+			// Must wait for cache sync, otherwise resource creation events will be missing if the resources are created
+			// in-between list and watch call of an informer. This is because fake clientset doesn't support watching with
+			// resourceVersion. A watcher of fake clientset only gets events that happen after the watcher is created.
+			controller.informerFactory.WaitForCacheSync(stopCh)
+			controller.crdInformerFactory.WaitForCacheSync(stopCh)
+			go controller.Run(stopCh)
+
+			var request *admv1.AdmissionRequest
+			if tc.oldSpec != nil && tc.newSpec != nil {
+				request = &admv1.AdmissionRequest{
+					Operation: admv1.Update,
+					OldObject: toRawExtension(tc.oldSpec),
+					Object:    toRawExtension(tc.newSpec),
+				}
+			} else if tc.newSpec != nil {
+				request = &admv1.AdmissionRequest{
+					Operation: admv1.Create,
+					Object:    toRawExtension(tc.newSpec),
+				}
+			} else {
+				t.Fatal("invalid test case")
+			}
```

Comment:
The input is under control, seems too verbose to handle impossible cases.

Reply:
When we add test cases, it may happen that the added test case is invalid. Reporting this may help debugging. I don't think three lines of validation would be too verbose.

Reply:
This is test for test, which is indeed verbose IMO. It may be only 3 lines in this test, but the same can apply to all tests. If we do such validation for all test cases, there could be thousands of lines and more. I have never seen a project adds tests for tests in practice.
To keep code clean, I would suggest only to assume error for uncontrollable cases, like when it makes an external call.

Reply:
I've deleted it. But still I have some opinions: sure, we cannot gurantee that code is 100% safe, or we will have tons of meta-test, meta-meta-test, and so on. However, this doesn't mean that trivial tests are meaningless. It looks like an "convention documentation" to me rather than an redundant test. It reminds me that such invalid test cases should never exist.